main/broadcasters.c

```diff
@@ -305,93 +305,166 @@ static broadcaster_ops_t eddystone_ops = {
-    sprintf(s, "%hhu", mijia_temp_hum->message_counter);
+    cb("MACAddress", _mactoa(mijia_header->mac), ctx);
+    sprintf(s, "%hhu", mijia_header->message_counter);
     cb("MessageCounter", s, ctx);
-    if (mijia_temp_hum->data_type == MIJIA_TEMP_HUM_DATA_TYPE_TEMP)
-    {
-        sprintf(s, "%.1f",
-            (int16_t)le16toh(*(uint16_t *)mijia_temp_hum->data) / 10.0);
-        cb("Temperature", s, ctx);
-    }
-    else if (mijia_temp_hum->data_type == MIJIA_TEMP_HUM_DATA_TYPE_HUM)
-    {
-        sprintf(s, "%.1f", le16toh(*(uint16_t *)mijia_temp_hum->data) / 10.0);
-        cb("Humidity", s, ctx);
-    }
-    else if (mijia_temp_hum->data_type == MIJIA_TEMP_HUM_DATA_TYPE_BATT)
-    {
-        sprintf(s, "%u", *mijia_temp_hum->data);
-        cb("BatteryLevel", s, ctx);
-    }
-    else if (mijia_temp_hum->data_type == MIJIA_TEMP_HUM_DATA_TYPE_TEMP_HUM)
+
+    if ((be16toh(mijia_header->frame_ctrl) & 0x4000) == 0)
+        return; /* No data */
+
+    mijia_data_entry_t *mijia_data_entry = (mijia_data_entry_t *)(
+        (uint8_t *)mijia_header + sizeof(mijia_header_t) +
+        ((be16toh(mijia_header->frame_ctrl) & 0x2000) ? 1 : 0) /* skip capability byte */);
+
+    while ((uint8_t*)mijia_data_entry - (uint8_t*)mijia_header < len)
     {
-        sprintf(s, "%.1f",
-            (int16_t)le16toh(*(uint16_t *)mijia_temp_hum->data) / 10.0);
-        cb("Temperature", s, ctx);
-        sprintf(s, "%.1f",
-            le16toh(*(uint16_t *)(mijia_temp_hum->data + 2)) / 10.0);
-        cb("Humidity", s, ctx);
+        if (mijia_data_entry->data_type == MIJIA_SENSOR_DATA_TYPE_TEMP)
+        {
+            sprintf(s, "%.1f",
+                (int16_t)le16toh(*(uint16_t *)mijia_data_entry->data) / 10.0);
+            cb("Temperature", s, ctx);
+        }
+        else if (mijia_data_entry->data_type == MIJIA_SENSOR_DATA_TYPE_SWITCH_TEMP)
+        {
+            sprintf(s, "%u", *mijia_data_entry->data);
+            cb("Switch", s, ctx);
+            sprintf(s, "%u", *(mijia_data_entry->data + 1));
+            cb("Temperature", s, ctx);
+        }
+        else if (mijia_data_entry->data_type == MIJIA_SENSOR_DATA_TYPE_HUM)
+        {
+            sprintf(s, "%.1f", le16toh(*(uint16_t *)mijia_data_entry->data) / 10.0);
+            cb("Humidity", s, ctx);
+        }
+        else if (mijia_data_entry->data_type == MIJIA_SENSOR_DATA_TYPE_MOIST ||
+                 mijia_data_entry->data_type == MIJIA_SENSOR_DATA_TYPE_MOIST2)
+        {
+            sprintf(s, "%u", *mijia_data_entry->data);
+            cb("Moisture", s, ctx);
+        }
+        else if (mijia_data_entry->data_type == MIJIA_SENSOR_DATA_TYPE_FDH)
+        {
+            sprintf(s, "%.1f", le16toh(*(uint16_t *)mijia_data_entry->data) / 100.0);
+            cb("Formaldehyde", s, ctx);
+        }
+        else if (mijia_data_entry->data_type == MIJIA_SENSOR_DATA_TYPE_LUM)
+        {
+            uint32_t val = 0;
+            memcpy((uint8_t*)&val, mijia_data_entry->data, 3);
```

Comment:
Copying 3 bytes to a 4-byte type? This might cause endianness issues.
I think it would be safer to address the data as bytes a do something like
`uint32_t val = mijia_data_entry->data[0] | (mijia_data_entry->data[1] << 8) .....`.
Or maybe even
`sprintf(s, "%u", le32toh(*(uint32_t *)mijia_data_entry->data) & 0xff000000);`